[SPARK-7961][SQL]Refactor SQLConf to display better error message

zsxwing opened the pull request:
1. Add `SQLConfEntry` to store the information about a configuration. For those configurations that cannot be found in `sql-programming-guide.md`, I left the doc as `<TODO>`.
2. Verify the value when setting a configuration if this is in SQLConf.
3. Use `SET -v` to display all public configurations.

zsxwing committed:
Refactor SQLConf to display better error message

1. Add `SQLConfEntry` to store the information about a configuration. For those configurations that cannot be found in `sql-programming-guide.md`, I left the doc as `<TODO>`.
2. Verify the value when setting a configuration if this is in SQLConf.
3. Add a command `conf` to display all public configurations.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34606 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34606/consoleFull) for   PR 6747 at commit [`326881d`](https://github.com/apache/spark/commit/326881d1b97a08ee45a46d12adcd19d86769fd1a).

zsxwing commented:
A simple example of the `conf` command:
```Scala
scala> sqlContext.sql("conf").collect().foreach(v => println(v.toSeq.mkString("\t")))
spark.sql.inMemoryColumnarStorage.batchSize	10000	Controls the size of batches for columnar caching.  Larger batch sizes can improve memory utilization and compression, but risk OOMs when caching data.
spark.sql.hive.verifyPartitionPath	true	<TODO>
spark.sql.retainGroupColumns	true	<TODO>
spark.sql.codegen	false	When true, code will be dynamically generated at runtime for expression evaluation in a specific query. For some queries with complicated expression this option can lead to significant speed-ups. However, for simple queries this can actually slow down query execution.
spark.sql.planner.sortMergeJoin	false	<TODO>
spark.sql.shuffle.partitions	200	Configures the number of partitions to use when shuffling data for joins or aggregations.
spark.sql.planner.externalSort	false	When true, performs sorts spilling to disk as needed otherwise sort each partition in memory.
spark.sql.broadcastTimeout	300	<TODO>
spark.sql.parquet.filterPushdown	false	Turn on Parquet filter pushdown optimization. This feature is turned off by default because of a known bug in Paruet 1.6.0rc3 (<a href="https://issues.apache.org/jira/browse/PARQUET-136">PARQUET-136</a>). However, if your table doesn't contain any nullable string or binary columns, it's still safe to turn this feature on.
spark.sql.sources.outputCommitterClass	<undefined>	
spark.sql.parquet.cacheMetadata	true	Turns on caching of Parquet schema metadata. Can speed up querying of static data.
spark.sql.unsafe.enabled	false	<TDDO>
spark.sql.useSerializer2	true	<TODO>
...
```

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34607 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34607/consoleFull) for   PR 6747 at commit [`f3c1b33`](https://github.com/apache/spark/commit/f3c1b3337fce260dfe8f37e18252cbb10741b1ca).

SparkQA commented:
[Test build #34606 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34606/console) for   PR 6747 at commit [`326881d`](https://github.com/apache/spark/commit/326881d1b97a08ee45a46d12adcd19d86769fd1a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #34607 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34607/console) for   PR 6747 at commit [`f3c1b33`](https://github.com/apache/spark/commit/f3c1b3337fce260dfe8f37e18252cbb10741b1ca).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -25,74 +25,283 @@ import scala.collection.JavaConversions._
+        }
+      }, _.toString, doc, isPublic)
+
+    def floatConf(
```
maybe remove the float version, and just use double version?

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -25,74 +25,283 @@ import scala.collection.JavaConversions._
+        isPublic: Boolean = true): SQLConfEntry[String] =
+      SQLConfEntry(key, defaultValue, v => v, v => v, doc, isPublic)
+
+    def seqConf[T](
```
how is this used?

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -131,44 +340,42 @@ private[sql] class SQLConf extends Serializable with CatalystConf {
   /** When true uses verifyPartitionPath to prune the path which is not exists. */
-  private[spark] def verifyPartitionPath =
-    getConf(HIVE_VERIFY_PARTITIONPATH, "true").toBoolean
+  private[spark] def verifyPartitionPath = getConf(HIVE_VERIFY_PARTITIONPATH)
```
can we add return type for all these functions?

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -25,74 +25,283 @@ import scala.collection.JavaConversions._
+
+  import SQLConfEntry._
+
+  val COMPRESS_CACHED = booleanConf("spark.sql.inMemoryColumnarStorage.compressed",
```
can we add a new line between confs to separate them?

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/SparkSQLParser.scala:
```diff
@@ -57,6 +57,7 @@ private[sql] class SparkSQLParser(fallback: String => LogicalPlan) extends Abstr
   protected val AS = Keyword("AS")
   protected val CACHE = Keyword("CACHE")
   protected val CLEAR = Keyword("CLEAR")
+  protected val CONF = Keyword("CONF")
```
conf is a bad name. let's figure out a better command for this ...

zsxwing replied:
I saw the following comments in the `SetCommand`:
```
    // Queries all key-value pairs that are set in the SQLConf of the sqlContext.
    // Notice that different from Hive, here "SET -v" is an alias of "SET".
    // (In Hive, "SET" returns all changed properties while "SET -v" returns all properties.)
    case Some(("-v", None)) | None =>
      sqlContext.getAllConfs.map { case (k, v) => Row(s"$k=$v") }.toSeq
```
How about using `SET -v` to display all configurations?

rxin replied:
that sounds good!

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -25,74 +25,283 @@ import scala.collection.JavaConversions._
+        entry
+      }
+
+    def intConf(
```
we can also remove the intConf, in favor of only longConf

zsxwing replied:
If we really want a `Int` conf, `intConf ` is better because it handles the overflow case. E.g.,
```
scala> Long.MaxValue.toString.toInt
java.lang.NumberFormatException: For input string: "9223372036854775807"
  at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
  at java.lang.Integer.parseInt(Integer.java:583)
  at java.lang.Integer.parseInt(Integer.java:615)
  at scala.collection.immutable.StringLike$class.toInt(StringLike.scala:272)
  at scala.collection.immutable.StringOps.toInt(StringOps.scala:30)
  ... 43 elided
```

rxin replied:
ok let's keep int then.

zsxwing committed:
Fix tests that use SQLConfEntry as a string

zsxwing committed:
Remove seqConf

zsxwing committed:
Add new lines between confs and return types

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34678 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34678/consoleFull) for   PR 6747 at commit [`99c9c16`](https://github.com/apache/spark/commit/99c9c162660826f2058d6eba10d5f8844664fda3).

zsxwing committed:
Remove the 'conf' command and use 'set -v' instead

zsxwing committed:
Remove floatConf

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34680 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34680/consoleFull) for   PR 6747 at commit [`8973ced`](https://github.com/apache/spark/commit/8973ced1b921fcf3ca330af51d1ee89f9de88792).

SparkQA commented:
[Test build #34678 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34678/console) for   PR 6747 at commit [`99c9c16`](https://github.com/apache/spark/commit/99c9c162660826f2058d6eba10d5f8844664fda3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #34680 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34680/console) for   PR 6747 at commit [`8973ced`](https://github.com/apache/spark/commit/8973ced1b921fcf3ca330af51d1ee89f9de88792).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

zsxwing committed:
Fix the compilation error

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34685 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34685/consoleFull) for   PR 6747 at commit [`6e47e56`](https://github.com/apache/spark/commit/6e47e5601d9e8f0061e9328f871fe481cdc0a293).

SparkQA commented:
[Test build #34685 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34685/console) for   PR 6747 at commit [`6e47e56`](https://github.com/apache/spark/commit/6e47e5601d9e8f0061e9328f871fe481cdc0a293).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

zsxwing committed:
Fix the test for 'set -v'

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34690 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34690/consoleFull) for   PR 6747 at commit [`4abd807`](https://github.com/apache/spark/commit/4abd80773ce9202f17b9eb205e2cf18158148328).

SparkQA commented:
[Test build #34690 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34690/console) for   PR 6747 at commit [`4abd807`](https://github.com/apache/spark/commit/4abd80773ce9202f17b9eb205e2cf18158148328).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

zsxwing commented:
Addressed all comments

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -25,74 +25,281 @@ import scala.collection.JavaConversions._
+      val valueConverter: String => T,
+      val stringConverter: T => String,
+      val doc: String,
+      val isPublic: Boolean) {
```
Can we add scaladoc for the semantics of these options.  In particular, `isPublic` isn't obvious to me.

zsxwing replied:
Added scaladoc for `SQLConfEntry`

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -25,74 +25,281 @@ import scala.collection.JavaConversions._
+    def defaultValueString: String = defaultValue.map(stringConverter).getOrElse("<undefined>")
+
+    override def toString: String = {
+      // Fail tests that use `SQLConfEntry` as a string.
```
What is the logic here?

marmbrus replied:
Basically my concern is not having a toString method can make things like debugging in the REPL much harder, though perhaps thats not an issue here.

zsxwing replied:
Sorry. It's a test code to help me debug. I just fixed it to a correct toString.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -25,74 +25,281 @@ import scala.collection.JavaConversions._
-  val THRIFTSERVER_POOL = "spark.sql.thriftserver.scheduler.pool"
-  val THRIFTSERVER_UI_STATEMENT_LIMIT = "spark.sql.thriftserver.ui.retainedStatements"
-  val THRIFTSERVER_UI_SESSION_LIMIT = "spark.sql.thriftserver.ui.retainedSessions"
+  val THRIFTSERVER_POOL = stringConf("spark.sql.thriftserver.scheduler.pool", isPublic = false)
```
This in particular is definitely a public API that people do use.

zsxwing replied:
Updated the `thriftserver` configurations

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -294,7 +501,31 @@ private[sql] class SQLConf extends Serializable with CatalystConf {
+   * Return the value of Spark SQL configuration property for the given key. If the key is not set
+   * yet, return `defaultValue` in [[SQLConfEntry]].
+   */
+  def getConf[T](entry: SQLConfEntry[T]): T = {
```
What about just having a `.value` method on the Conf entry itself?

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -219,74 +422,78 @@ private[sql] class SQLConf extends Serializable with CatalystConf {
-  /** Set the given Spark SQL configuration property. */
-  def setConf(key: String, value: String): Unit = {
+  /** Set the given Spark SQL configuration property using a `string` value. */
+  def setRawConf(key: String, value: String): Unit = {
```
Why `Raw`? Maybe, `setConfString`?

zsxwing replied:
Changed it to `setConfString` and also changed `getRawConf` to `getConfString`

marmbrus commented:
Thanks for working on this much needed cleanup!  A few comments (some of which were preexisting):

 - The schema for set commands is wrong (we need some column name)
```
scala> sql("SET -v")
res4: org.apache.spark.sql.DataFrame = [: string]
```

- Related to the above, I think we should consider breaking out the parts of commands like `SET` and `SET -v` instead of just returning a string.

 - When the option has a default, I think we should show that instead of `<undefined>`

```
scala> sql("SET spark.sql.autoBroadcastJoinThreshold").collect()
res1: Array[org.apache.spark.sql.Row] = Array([spark.sql.autoBroadcastJoinThreshold=<undefined>])
```

- It seems a bunch of boilerplate could be removed and we could get some more typesafety by making the conf entries inter classes of SQLConf with methods like `.value` and `.set(...)`.  This would avoid having to have an entry and a `def` for each config option.  Perhaps I'm missing something though.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -25,74 +25,281 @@ import scala.collection.JavaConversions._
+    defaultValue = Some(true),
+    doc = "Turns on caching of Parquet schema metadata. Can speed up querying of static data.")
+
+  val PARQUET_COMPRESSION = stringConf("spark.sql.parquet.compression.codec",
```
Could we also have an entry type for configs like these that validates the option is in some list?

zsxwing replied:
Added `enumConf` and used it for `PARQUET_COMPRESSION`

zsxwing committed:
setRawString => setConfString, getRawString => getConfString

zsxwing committed:
Add enumConf

zsxwing committed:
Add doc for SQLConfEntry and fix 'toString'

zsxwing committed:
Fix the configurations about HiveThriftServer

zsxwing committed:
Merge branch 'master' into sqlconf

zsxwing committed:
Add schema to SetCommand

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34763 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34763/consoleFull) for   PR 6747 at commit [`037b1db`](https://github.com/apache/spark/commit/037b1dbb33f1a2abbce29867118cb001d0d7deb6).

SparkQA commented:
[Test build #34763 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34763/console) for   PR 6747 at commit [`037b1db`](https://github.com/apache/spark/commit/037b1dbb33f1a2abbce29867118cb001d0d7deb6).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

zsxwing committed:
getConf will return the default value if a config is not set

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -25,74 +25,314 @@ import scala.collection.JavaConversions._
+          stringConverter: T => String,
+          doc: String,
+          isPublic: Boolean): SQLConfEntry[T] =
+      sqlConfEntries.synchronized {
```
Is `sqlConfEntries` already a synchronized map?

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34767 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34767/consoleFull) for   PR 6747 at commit [`a2f4add`](https://github.com/apache/spark/commit/a2f4add15f3707752b14f1b808ac4c9f00efac8d).

zsxwing commented:
> The schema for set commands is wrong (we need some column name)

Added schemas for all set commands and split results into columns. For the configuration that has a default value, if it's not set , `set <configuration>` will show the default value.

> It seems a bunch of boilerplate could be removed and we could get some more typesafety by making the conf entries inter classes of SQLConf with methods like .value and .set(...). This would avoid having to have an entry and a def for each config option. Perhaps I'm missing something though.

Sounds good. I will try it.

zsxwing commented:
> It seems a bunch of boilerplate could be removed and we could get some more typesafety by making the conf entries inter classes of SQLConf with methods like .value and .set(...). This would avoid having to have an entry and a def for each config option. Perhaps I'm missing something though.

One disadvantage is the size of a  `SQLConf` instance will increase a lot since all SQLConfEntries are put into it.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -304,6 +574,16 @@ private[sql] class SQLConf extends Serializable with CatalystConf {
+    }.toSeq
+  }
+
   private[spark] def unsetConf(key: String) {
```
Adding a `def unsetConf[T](entry: SQLConfEntry[T])` here can be convenient.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -219,82 +455,116 @@ private[sql] class SQLConf extends Serializable with CatalystConf {
+      // Try to use the default value
+      Option(sqlConfEntries.get(key)).map(_.defaultValueString)
+      }.
+      getOrElse(throw new NoSuchElementException(key))
```
Indentations and newline are off here.

zsxwing committed:
Add unsetConf(SQLConfEntry) and fix the code style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34768 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34768/consoleFull) for   PR 6747 at commit [`3c5f03e`](https://github.com/apache/spark/commit/3c5f03e592f505fd1aa956ad28a9aede4b16cc20).

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/execution/commands.scala:
```diff
@@ -76,47 +78,93 @@ private[sql] case class ExecutedCommand(cmd: RunnableCommand) extends SparkPlan
 case class SetCommand(
-    kv: Option[(String, Option[String])],
-    override val output: Seq[Attribute])
+    kv: Option[(String, Option[String])])
```
Nit: Seems that we can remove the newline here.

liancheng commented:
One of the things that I envy Hive is `ConfVars`, and now we finally have our own version :)

Some follow-ups I can think of (but not necessarily to be done in this PR):

- Add generic version of `setConf`, `getConf`, `unsetConf` to `SQLContext` (note that `HiveContext.setConf` overrides `SQLContext.setConf` and also manipulates `HiveConf` there)
- Replace all occurrences of `sqlContext.setConf/getConf/unsetConf` to the corresponding generic version

SparkQA commented:
[Test build #34767 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34767/console) for   PR 6747 at commit [`a2f4add`](https://github.com/apache/spark/commit/a2f4add15f3707752b14f1b808ac4c9f00efac8d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #34768 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34768/console) for   PR 6747 at commit [`3c5f03e`](https://github.com/apache/spark/commit/3c5f03e592f505fd1aa956ad28a9aede4b16cc20).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

liancheng commented:
Another thought is that, we probably also want to migrate data sources options to this new API.

@marmbrus How do you think?

zsxwing committed:
Fix the code style and tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34775 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34775/consoleFull) for   PR 6747 at commit [`cf950c1`](https://github.com/apache/spark/commit/cf950c1c5aea79ccb5899da5a72f5b14b23bbd99).

SparkQA commented:
[Test build #34775 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34775/console) for   PR 6747 at commit [`cf950c1`](https://github.com/apache/spark/commit/cf950c1c5aea79ccb5899da5a72f5b14b23bbd99).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SetCommand(kv: Option[(String, Option[String])]) extends RunnableCommand with Logging `
  * `  case class SetInFilter[T <: Comparable[T]](`

AmplabJenkins commented:
Merged build finished. Test FAILed.

zsxwing committed:
Fix the unit tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test PASSed.

zsxwing committed:
Merge branch 'master' into sqlconf

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34982 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34982/consoleFull) for   PR 6747 at commit [`e014f53`](https://github.com/apache/spark/commit/e014f5361feb98929b98149cd181fef1c376abe4).

SparkQA commented:
[Test build #34982 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34982/console) for   PR 6747 at commit [`e014f53`](https://github.com/apache/spark/commit/e014f5361feb98929b98149cd181fef1c376abe4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class IsNull(child: Expression) extends UnaryExpression with Predicate `
  * `case class IsNotNull(child: Expression) extends UnaryExpression with Predicate `
  * `case class SetCommand(kv: Option[(String, Option[String])]) extends RunnableCommand with Logging `

AmplabJenkins commented:
Merged build finished. Test PASSed.

zsxwing committed:
Add getConf, setConf to SQLContext and HiveContext

zsxwing committed:
Use SQLConfEntry in HiveContext

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35047 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35047/consoleFull) for   PR 6747 at commit [`7d09bad`](https://github.com/apache/spark/commit/7d09bad23a7ac25c90735079b01984fa307a6f73).

SparkQA commented:
[Test build #35047 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35047/console) for   PR 6747 at commit [`7d09bad`](https://github.com/apache/spark/commit/7d09bad23a7ac25c90735079b01984fa307a6f73).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SetCommand(kv: Option[(String, Option[String])]) extends RunnableCommand with Logging `

AmplabJenkins commented:
Merged build finished. Test PASSed.

rxin commented:
Alright merging this. Thanks.
